engine/crates/engine-v2/query-planning/src/tests/entities.rs

```diff
@@ -0,0 +1,97 @@
+use crate::{
+    tests::{read_schema, TestOperation},
+    OperationGraph,
+};
+
+const SCHEMA: &str = r###"
+enum join__Graph {
+  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
+  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
+  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
+  REVIEW @join__graph(name: "review", url: "http://localhost:4200/shared-root/review")
+}
+
+type Product
+  @join__type(graph: CATEGORY, key: "id")
+  @join__type(graph: NAME, key: "id")
+  @join__type(graph: PRICE, key: "id")
+  @join__type(graph: REVIEW, key: "id")
+{
+  id: ID!
+  reviews: [Review] @join__field(graph: REVIEW)
+  category: String @join__field(graph: CATEGORY)
+  name: String @join__field(graph: NAME)
+  price: Float @join__field(graph: PRICE)
+}
+
+type Review @join__type(graph: REVIEW)
+{
+  stars: Int @join__field(graph: REVIEW)
+}
+
+type Query
+  @join__type(graph: CATEGORY)
+  @join__type(graph: NAME)
+  @join__type(graph: PRICE)
+{
+  product: Product
+  products: [Product]
+}
+"###;
+
+#[test]
+fn two_fields() {
+    let schema = read_schema(SCHEMA);
+    let mut operation = TestOperation::bind(
+        &schema,
+        r#"
+        query {
+          products {
+            price
+            category
+          }
+        }
+    "#,
+    );
+
+    let graph = OperationGraph::new(&schema, &mut operation);
+    insta::assert_snapshot!("two_fields-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());
+}
+
+#[test]
+fn single_field() {
+    let schema = read_schema(SCHEMA);
+    let mut operation = TestOperation::bind(
+        &schema,
+        r#"
+        query {
+          products {
+            price
+          }
+        }
+    "#,
+    );
+
+    let graph = OperationGraph::new(&schema, &mut operation);
+    insta::assert_snapshot!("single_field-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());
+}
+
+#[test]
+fn nested_join() {
+    let schema = read_schema(SCHEMA);
+    let mut operation = TestOperation::bind(
+        &schema,
+        r#"
+        query {
+          products {
+            reviews {
+              stars
+            }
+          }
+        }
+    "#,
+    );
+
+    let graph = OperationGraph::new(&schema, &mut operation);
+    insta::assert_snapshot!("nested_join-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());
```

Comment:
I'm not sure do I like these tests. You can't really see what should happen, just that these two things must be the same. Maybe the output is just too long to paste here, but for the reader it immediately tells what is expected, and also how to construct the next test.

Reply:
It's a snapshot with an output as follows. It's unreadable as is, but it's easy to understand the difference. The third argument is the description of the snapshot. It gets printed whenever the test fails. This description is the readable format when using graphviz to show the graph.

```text
digraph {
    0 [ root]
    1 [ stars]
    2 [ reviews]
    3 [ products]
    4 [ Root#category]
    5 [ products@Root#category]
    6 [ Root#name]
    7 [ products@Root#name]
    8 [ Root#price]
    9 [ products@Root#price]
    10 [ FedEntity#review]
    11 [ reviews@FedEntity#review]
    12 [ stars@FedEntity#review]
    13 [ *id]
    14 [ id@Root#category]
    15 [ id@Root#name]
    16 [ id@Root#price]
    0 -> 3 [ label = "Field" ]
    0 -> 4 [ label = "CreateChildResolver(1)" ]
    0 -> 4 [ label = "HasChildResolver" ]
    4 -> 5 [ label = "CanProvide(0)" ]
    5 -> 3 [ label = "Provides" ]
    0 -> 6 [ label = "CreateChildResolver(1)" ]
    0 -> 6 [ label = "HasChildResolver" ]
    6 -> 7 [ label = "CanProvide(0)" ]
    7 -> 3 [ label = "Provides" ]
    0 -> 8 [ label = "CreateChildResolver(1)" ]
    0 -> 8 [ label = "HasChildResolver" ]
    8 -> 9 [ label = "CanProvide(0)" ]
    9 -> 3 [ label = "Provides" ]
    3 -> 2 [ label = "Field" ]
    9 -> 10 [ label = "CreateChildResolver(1)" ]
    3 -> 10 [ label = "HasChildResolver" ]
    10 -> 11 [ label = "CanProvide(0)" ]
    11 -> 2 [ label = "Provides" ]
    2 -> 1 [ label = "Field" ]
    11 -> 12 [ label = "CanProvide(0)" ]
    12 -> 1 [ label = "Provides" ]
    7 -> 10 [ label = "CreateChildResolver(1)" ]
    5 -> 10 [ label = "CreateChildResolver(1)" ]
    10 -> 13 [ label = "Requires" ]
    3 -> 13 [ label = "Field" ]
    5 -> 14 [ label = "CanProvide(0)" ]
    14 -> 13 [ label = "Provides" ]
    7 -> 15 [ label = "CanProvide(0)" ]
    15 -> 13 [ label = "Provides" ]
    9 -> 16 [ label = "CanProvide(0)" ]
    16 -> 13 [ label = "Provides" ]
}```